lib/fluent/plugin/in_s3.rb

```diff
@@ -0,0 +1,285 @@
+      @bucket = @s3.bucket(@s3_bucket)
+
+      raise "#{@bucket.name} is not found." unless @bucket.exists?
+
+      check_apikeys if @check_apikey_on_start
+
+      @running = true
+      @thread = Thread.new(&method(:run))
+    end
+
+    def shutdown
+      @running = false
+      @thread.join
+      super
+    end
+
+    private
+
+    def run
+      options = {}
+      options[:wait_time_seconds] = @sqs.wait_time_seconds
+      options[:skip_delete] = @sqs.skip_delete
+      @poller.before_request do |stats|
+        throw :stop_polling unless @running
+      end
+      @poller.poll(options) do |message|
+        begin
+          body = Yajl.load(message.body)
+          next unless body["Records"] # skip test queue
+
+          process(body)
+        rescue => e
+          log.warn "#{e.message}\n#{e.backtrace.join("\n")}"
+          @running = false
+          throw :skip_delete
+        end
+      end
+    end
+
+    def setup_credentials
+      options = {}
+      credentials_options = {}
+      case
+      when @aws_key_id && @aws_sec_key
+        options[:access_key_id] = @aws_key_id
+        options[:secret_access_key] = @aws_sec_key
+      when @assume_role_credentials
+        c = @assume_role_credentials
+        credentials_options[:role_arn] = c.role_arn
+        credentials_options[:role_session_name] = c.role_session_name
+        credentials_options[:policy] = c.policy if c.policy
+        credentials_options[:duration_seconds] = c.duration_seconds if c.duration_seconds
+        credentials_options[:external_id] = c.external_id if c.external_id
+        options[:credentials] = Aws::AssumeRoleCredentials.new(credentials_options)
+      when @instance_profile_credentials
+        c = @instance_profile_credentials
+        credentials_options[:retries] = c.retries if c.retries
+        credentials_options[:ip_address] = c.ip_address if c.ip_address
+        credentials_options[:port] = c.port if c.port
+        credentials_options[:http_open_timeout] = c.http_open_timeout if c.http_open_timeout
+        credentials_options[:http_read_timeout] = c.http_read_timeout if c.http_read_timeout
+        options[:credentials] = Aws::InstanceProfileCredentials.new(credentials_options)
+      when @shared_credentials
+        c = @shared_credentials
+        credentials_options[:path] = c.path if c.path
+        credentials_options[:profile_name] = c.profile_name if c.profile_name
+        options[:credentials] = Aws::SharedCredentials.new(credentials_options)
+      else
+        # Use default credentials
+        # See http://docs.aws.amazon.com/sdkforruby/api/Aws/S3/Client.html
+      end
+      options
+    end
+
+    def create_s3_client
+      options = setup_credentials
+      options[:region] = @s3_region if @s3_region
+      options[:proxy_uri] = @proxy_uri if @proxy_uri
+      options[:s3_server_side_encryption] = @use_server_side_encryption.to_sym if @use_server_side_encryption
```

Comment:
Maybe, no.
When someone has read  access to encrypted object (by SSE) on S3 bucket, he can download decrypted object from S3 bucket.
Is that right?